ext/symengine/ruby_matrix.c

```diff
@@ -0,0 +1,123 @@
+#include "ruby_matrix.h"
+
+void cmatrix_dense_free(void *ptr)
+{
+    CDenseMatrix *mat_ptr = ptr;
+    dense_matrix_free(mat_ptr);
+}
+
+VALUE cmatrix_dense_alloc(VALUE klass)
+{
+    CDenseMatrix *mat_ptr = dense_matrix_new();
+    return Data_Wrap_Struct(klass, NULL, cmatrix_dense_free, mat_ptr);
+}
+
+VALUE cmatrix_to_str(VALUE self)
+{
+    CDenseMatrix *this;
+    char *str_ptr;
+    VALUE result;
+
+    Data_Get_Struct(self, CDenseMatrix, this);
+
+    str_ptr = matrix_str(this);
+    result = rb_str_new_cstr(str_ptr);
+    basic_str_free(str_ptr);
+
+    return result;
+}
+
+
+VALUE cmatrix_dense_init(VALUE self, VALUE args)
+{
+    int argc = RARRAY_LEN(args);
+    CDenseMatrix *this;
+    Data_Get_Struct(self, CDenseMatrix, this);
+    
+    if (argc == 0) {
+    
+    // SymEngine::DenseMatrix()
+        dense_matrix(this);
+        
+    } else if (argc == 1) {
+    // SymEngine::DenseMatrix(SymEngine::DenseMatrix) OR
+    // SymEngine::DenseMatrix(NMatrix) OR
+    // SymEngine::DenseMatrix(Array)
+    
+        VALUE operand = rb_ary_shift(args);
+        char *s = rb_obj_classname(operand);
+        
+        if(strcmp(s, "SymEngine::DenseMatrix") == 0) {
```

Comment:
Will format all code with clang-format. Please disregard formating issues for now.